other/n_body_simulation.py

```diff
@@ -0,0 +1,262 @@
+        """
+        for body1 in self.bodies:
+            force_x = 0.0
+            force_y = 0.0
+            for body2 in self.bodies:
+                if body1 != body2:
+                    dif_x = body2.position_x - body1.position_x
+                    dif_y = body2.position_y - body1.position_y
+
+                    # Calculation of the distance using Pythagoras's theorem
+                    # Extra factor due to the softening technique
+                    distance = (dif_x ** 2 + dif_y ** 2 + self.softening_factor) ** (
+                        1 / 2
+                    )
+
+                    # Newton's law of universal gravitation.
+                    force_x += (
+                        self.gravitation_constant * body2.mass * dif_x / distance ** 3
+                    )
+                    force_y += (
+                        self.gravitation_constant * body2.mass * dif_y / distance ** 3
+                    )
+
+            # Update the body's velocity once all the force components have been added
+            body1.update_velocity(force_x, force_y, delta_time * self.time_factor)
+
+        # Update the positions only after all the velocities have been updated
+        for body in self.bodies:
+            body.update_position(delta_time * self.time_factor)
+
+
+def plot(
+    title: str,
+    body_system: BodySystem,
+    x_start: float = -1,
+    x_end: float = 1,
+    y_start: float = -1,
+    y_end: float = 1,
+) -> None:
+    """
+    Utility function to plot how the given body-system evolves over time.
+    No doctest provided since this function does not have a return value.
+    """
+
+    INTERVAL = 20  # Frame rate of the animation
+    DELTA_TIME = INTERVAL / 1000  # Time between time steps in seconds
+
+    fig = plt.figure()
+    fig.canvas.set_window_title(title)
+
+    # Set section to be plotted
+    ax = plt.axes(xlim=(x_start, x_end), ylim=(y_start, y_end))
+
+    # Each body is drawn as a patch by the plt-function
+    patches = []
+    for body in body_system.bodies:
+        patches.append(
+            plt.Circle((body.position_x, body.position_y), body.size, fc=body.color)
+        )
+
+    # Function called once at the start of the animation
+    def init() -> list[patches.Circle]:
+        axes = plt.gca()
+        axes.set_aspect("equal")
+
+        for patch in patches:
+            ax.add_patch(patch)
+        return patches
+
+    # Function called at each step of the animation
+    def update(frame: int) -> list[patches.Circle]:
+        # Update the positions of the bodies
+        body_system.update_system(DELTA_TIME)
+
+        # Update the positions of the patches
+        for patch, body in zip(patches, body_system.bodies):
+            patch.center = (body.position_x, body.position_y)
+        return patches
+
+    anim = animation.FuncAnimation(  # noqa: F841
+        fig, update, init_func=init, interval=INTERVAL, blit=True
+    )
+
+    plt.show()
+
+
+if __name__ == "__main__":

```

Comment:
Please break this long collection of code into three separate functions.

```suggestion
if __name__ == "__main__":
    example_1()
    example_2()
    example_3()
```